File: python/lesson 7/binary_search_tree.py
# https://leetcode.com/problems/search-in-a-binary-search-tree/
# https://leetcode.com/problems/insert-into-a-binary-search-tree/
# https://leetcode.com/problems/delete-node-in-a-bst/

class BinarySearchTree:
    def __init__(self):
        self.root = None
        
    def get(self, val: int):
        self.get_value(self, self.root, val)
        
    def get_value(self, node: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if node == None: return None
        
        if val < node.val:
            return self.get_value(node.left, val)
        elif val > node.val:
            return self.get_value(node.right, val)
        else:
            return node 
        
    def put(self, val: int):
        self.root = self.put_value(self, self.root, val)
    
    def put_value(self, node: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if node == None: return TreeNode(val)
        
        if val < node.val:
            node.left = self.put_value(node.left, val)
        elif val > node.val:
            node.right = self.put_value(node.right, val)
        
        return node 
    
    def delete(self, val: int):
        self.root = self.delete_value(self.root, val)
    
    def delete_value(self, node: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if node == None: return None
        
        if val < node.val:
            node.left = self.delete_value(node.left, val)
        elif val > node.val:
            node.right = self.delete_value(node.right, val)
        else:
            if node.right == None:
                return node.left
            if node.left == None:
                return node.right
            temp = node
            node = self.min(temp.right)
            node.right = self.delete_min(temp.right)
            node.left = temp.left
        return node
    
    def min(self, node: Optional[TreeNode]) -> Optional[TreeNode]:
        if node.left == None:
            return node
        return self.min(node.left)
    
    def delete_min(self, node: Optional[TreeNode]) -> Optional[TreeNode]:
        if node.left == None:
            return node.right
        node.left = self.delete_min(node.left)
        return node

File: kotlin/lesson 7/BinarySearchTree.kt
// https://leetcode.com/problems/search-in-a-binary-search-tree/
// https://leetcode.com/problems/insert-into-a-binary-search-tree/
// https://leetcode.com/problems/delete-node-in-a-bst/

class BinarySearchTree {
    var root: TreeNode? = null

    fun get(`val`: Int) = get(root, `val`)

    private fun get(node: TreeNode?, `val`: Int): TreeNode? {
        if (node == null) return null
         
        return if (`val` < node.`val`)
                get(node.left, `val`)
            else if (`val` > node.`val`)
                get(node.right, `val`)
            else 
                node
    }

    fun put(`val`: Int) {
        this.root = put(root, `val`)
    }

    private fun put(node: TreeNode?, `val`: Int): TreeNode? {
        if (node == null) return TreeNode(`val`)

        if (`val` < node.`val`)
            node.left = put(node.left, `val`)
        else if (`val` > node.`val`)
            node.right = put(node.right, `val`)

        return node
    }

    fun delete(`val`: Int) {
        this.root = delete(root, `val`)
    }

    private fun delete(node: TreeNode?, `val`: Int): TreeNode? {
        if (node == null) return null
        
        if (`val` < node.`val`)
            node.left = delete(node.left, `val`)
        else if (`val` > node.`val`)
            node.right = delete(node.right, `val`)
        else {
            if (node.right == null)
                return node.left
            else if (node.left == null)
                return node.right
            else {
                var min = min(node.right)
                min.right = deleteMin(node.right)
                min.left = node.left
                return min
            }
        }
        return node
    }
    
    private fun min(node: TreeNode): TreeNode {
        return if (node.left == null) node else min(node.left)
    } 
    
    private fun deleteMin(node: TreeNode): TreeNode? {
        if (node.left == null) return node.right
        node.left = deleteMin(node.left)
        return node
    }
}
